[MNG-7182] Use a pull parser during the build/consumer transformation

gnodet opened the pull request:
Following this checklist to help us incorporate your 
contribution quickly and easily:

 - [x] Make sure there is a [JIRA issue](https://issues.apache.org/jira/browse/MNG) filed 
       for the change (usually before you start working on it).  Trivial changes like typos do not 
       require a JIRA issue.  Your pull request should address just this issue, without 
       pulling in other changes.
 - [x] Each commit in the pull request should have a meaningful subject line and body.
 - [x] Format the pull request title like `[MNG-XXX] - Fixes bug in ApproximateQuantiles`,
       where you replace `MNG-XXX` with the appropriate JIRA issue. Best practice
       is to use the JIRA issue title in the pull request title and in the first line of the 
       commit message.
 - [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
 - [x] Run `mvn clean verify` to make sure basic checks pass. A more thorough check will 
       be performed on your pull request automatically.
 - [x] You have run the [Core IT][core-its] successfully.

If your pull request is about ~20 lines of code you don't need to sign an
[Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf) if you are unsure
please ask on the developers list.

To make clear that you license your contribution under 
the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0)
you have to acknowledge this by using the following check-box.

 - [x] I hereby declare this contribution to be licenced under the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0)

 - [x] In any other case, please file an [Apache Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

[core-its]: https://maven.apache.org/core-its/core-it-suite/

michael-o commented:
I think this needs a JIRA issue.

gnodet commented:
There's a single test failing on windows + jdk 1.8.  Does that ring any bell @rfscholte ?
  https://github.com/apache/maven/pull/486/checks?check_run_id=3016896271#step:8:6012

michael-o commented:
> 
> 
> There's a single test failing on windows + jdk 1.8. Does that ring any bell @rfscholte ?
> https://github.com/apache/maven/pull/486/checks?check_run_id=3016896271#step:8:6012

Bad code, concurrent write access, Windows file locking.

rfscholte commented:
That would be my guess too

rfscholte commented:
It is a bit weird that this IT is using an external plugin for verification. First step would be to update the version of the flatten-maven-plugin. Maybe in the future we need to reconsider rewriting this IT.

MartinKanters commented on maven-model-transform/src/main/java/org/apache/maven/model/transform/BuildToRawPomXMLFilterFactory.java:
```diff
@@ -44,74 +36,40 @@
-        XMLReader xmlReader = Factories.newXMLReader();
-        xmlReader.setFeature( "http://xml.org/sax/features/namespaces", true );
-        return xmlReader;
+        return parser;
```
The `parser` param is overwritten a couple of times, which can result into unexpected results for the calling code. I think creating a new parser variable at the start of the method which would be returned in the end is more clear.

MartinKanters commented on maven-model-transform/src/main/java/org/apache/maven/model/transform/FastForwardFilter.java:
```diff
@@ -19,23 +19,23 @@
 
 /**
  * This filter will skip all following filters and write directly to the output.
  * Should be used in case of a DOM that should not be effected by other filters, even though the elements match
```
Let's finish the sentence with a dot here, while we are at it.

MartinKanters commented on maven-model-transform/src/main/java/org/apache/maven/model/transform/FastForwardFilter.java:
```diff
@@ -53,75 +53,50 @@
-        else
+        else if ( xmlPullParser.getEventType() == END_TAG )
         {
+            if ( --domDepth == 0 )
```
Small one, I prefer splitting up decreasing the variable and having this condition. It might be just me, but my brain takes a couple of extra cycles remembering whether the return value is before or after the operation.

MartinKanters commented on maven-model-transform/src/main/java/org/apache/maven/model/transform/RawToConsumerPomXMLFilterFactory.java:
```diff
@@ -41,20 +37,19 @@ public RawToConsumerPomXMLFilterFactory( BuildToRawPomXMLFilterFactory buildPomX
+        parser = new RelativePathXMLFilter( parser );
 
-        return new RawToConsumerPomXMLFilter( filter );
+        return parser;
```
Same comment as above, let's not alter the parameter

MartinKanters commented on maven-model-transform/src/main/java/org/apache/maven/model/transform/RelativePathXMLFilter.java:
```diff
@@ -19,90 +19,60 @@
+            }
+            else
+            {
+                if ( skip )
```
This if can combined with the else on line 62

MartinKanters commented on maven-model-transform/src/test/java/org/apache/maven/model/transform/ConsumerPomXMLFilterTest.java:
```diff
@@ -82,10 +69,9 @@ protected AbstractSAXFilter getFilter( Consumer<LexicalHandler> lexicalHandlerCo
-        return filter;
+        parser = new RawToConsumerPomXMLFilterFactory( buildPomXMLFilterFactory )
+                        .get( parser, Paths.get( "pom.xml" ) );
+        return parser;
```
Same comment here, but in this case you can directly return the statement: `return new RawToConsumerPomXMLFilterFactory(...`

gnodet replied:
I really don't get the problem, but once again, I've added an intermediary variable.
Some people also prefer not returning instances created directly to ease debugging.

MartinKanters commented on maven-model-transform/src/test/java/org/apache/maven/model/transform/ModulesXMLFilterTest.java:
```diff
@@ -19,23 +19,19 @@
 
     @Override
-    protected ModulesXMLFilter getFilter( Consumer<LexicalHandler> lexicalHandlerConsumer )
+    protected ModulesXMLFilter getFilter(XmlPullParser parser)
```
Checkstyle should not like this, the param should be prefixed and suffixed with spaces

MartinKanters commented on maven-core/src/main/java/org/apache/maven/internal/aether/ConsumerModelSourceTransformer.java:
```diff
@@ -23,91 +23,30 @@
+        XmlStreamReader reader = ReaderFactory.newXmlReader( Files.newInputStream( pomFile ) );
+        XmlPullParser parser = new MXParser( EntityReplacementMap.defaultEntityReplacementMap );
+        parser.setInput( reader );
+        parser = new RawToConsumerPomXMLFilterFactory( new DefaultBuildPomXMLFilterFactory( context, true ) )
```
Instead of overwriting the old parser, I think it's clearer if we create a new variable for it.

MartinKanters reviewed (commented):
The business logic is new to me and I'm not too experienced with XML transformation. I didn't see any red flags, but perhaps @rfscholte for example can take a better look at that? My review is more focused on small coding things. Nice job on the performance boost!

gnodet commented:
@rfscholte could you please review, I think that one is important to include in master asap to avoid the performance loss in the first alpha.
There's still this small problem on windows + jdk8 (file lock still in place) which I've worked around by using a fork + timeout in the tests, but It does not appear on any other combination and I've been unable to find a place where a file lock would be kept.  In addition, I haven't been able to reproduce it in a real environment.

rfscholte commented:
Just ran the branch on Jenkins, it is unstable: https://ci-maven.apache.org/blue/organizations/jenkins/Maven%2Fmaven-box%2Fmaven/detail/MNG-7182/1/tests
One thing I notice is that it looks like the linebreaks around the license are better compared to what I could do. Please also provide a PR for that.

gnodet commented:
> Just ran the branch on Jenkins, it is unstable: https://ci-maven.apache.org/blue/organizations/jenkins/Maven%2Fmaven-box%2Fmaven/detail/MNG-7182/1/tests
> One thing I notice is that it looks like the linebreaks around the license are better compared to what I could do. Please also provide a PR for that.

Yes, it took me a bit of time to get that right ;-)
I'll have a look at those tests, I thought I did modify them to pass.

gnodet commented:
> > Just ran the branch on Jenkins, it is unstable: https://ci-maven.apache.org/blue/organizations/jenkins/Maven%2Fmaven-box%2Fmaven/detail/MNG-7182/1/tests
> > One thing I notice is that it looks like the linebreaks around the license are better compared to what I could do. Please also provide a PR for that.
> 
> Yes, it took me a bit of time to get that right ;-)
> I'll have a look at those tests, I thought I did modify them to pass.

I fixed and squashed the PRs.

cstamas commented:
Re `Caused by: java.lang.NoSuchMethodError: org.apache.maven.plugin.PluginParameterExpressionEvaluator.<init>` see here https://github.com/apache/maven/pull/562#discussion_r720683592
Do you have some ancient enforcer? Or comment is wrong and there is something else (ancient) using this ctor?

cstamas commented:
> Re `Caused by: java.lang.NoSuchMethodError: org.apache.maven.plugin.PluginParameterExpressionEvaluator.<init>` see here [#562 (comment)](https://github.com/apache/maven/pull/562#discussion_r720683592) Do you have some ancient enforcer? Or comment is wrong and there is something else (ancient) using this ctor?

Hm, but something is wrong, it WAS ancient and is already upped https://github.com/apache/maven-integration-testing/pull/115

cstamas commented:
> Hm, but something is wrong, it WAS ancient and is already upped [apache/maven-integration-testing#115](https://github.com/apache/maven-integration-testing/pull/115)

Eh, update your fork please https://github.com/gnodet/maven-integration-testing

gnodet commented:
> > Hm, but something is wrong, it WAS ancient and is already upped [apache/maven-integration-testing#115](https://github.com/apache/maven-integration-testing/pull/115)
> 
> Eh, update your fork please https://github.com/gnodet/maven-integration-testing (master on apache/maven-integration-testing contains fixes for you: upped enforcer)

Thx, I've rebased both PRs on top of master, which looks much better indeed !

gnodet commented:
I'd like to get that in `master` to avoid the big performance loss.  Anyone could take a look ?

rmannibucau reviewed (approved):
(LGTM, the very minor note can be to write a one sentence description in the ticket since it lacks some context if anyone reviews it later ;))

hboutemy commented:
@rfscholte do you see any drawback to using plexus-utils' XPP3?

jira-importer commented:
Resolve #8266
